Orbit: Send telemetry for the UI

MaEtUgR opened the pull request:
@potaito Thanks for your help!
@DonLakeFlyer You can use this branch for UI testing.

**Reference**
- Base branch for this to enable Orbit, needs to be merged first: #10907
- `ORBIT_EXECUTION_STATUS` MAVLink telemetry message: https://github.com/mavlink/mavlink/pull/1006 

**Test data / coverage**
SITL testing, looking at "Analyze" widget in latest daily QGC to see the `ORBIT_EXECUTION_STATUS` message output.

**Describe problem solved by the proposed pull request**
The pilot commands an orbit but does not see it's status during execution. He has to remember what he commanded and any change in the execution parameters are not reflected in the UI. Currently the orbit can be influenced by RC sticks and in the future hopefully also by moving the center target dynamically e.g. using a beacon.

**Describe your preferred solution**
Fill the MAVLink message for the vehicle to report the current status, parameters and therefore planned trajectory to the ground control station while flying an orbit to allow the UI to show a live visualization for the pilot. **The message is only sent out while the orbit is being executed.**

**Describe possible alternatives**
Sending back the command which anyways has almost the same fields with a flag that it's telemetry, discussed with @DonLakeFlyer to be a terrible idea.

**Additional context**
<img width="990" alt="telemetry screenshot" src="https://user-images.githubusercontent.com/4668506/49519018-5176dc00-f8a0-11e8-9c1a-ec21e63c510d.PNG">

MaEtUgR commented:
Frame coordinate tranformation of the center point is missing, working on that.

MaEtUgR commented:
I think it should be fixed now.

DonLakeFlyer commented:
When I go into Orbit it reports HEARTBEAT.custom_mode == 0 which is incorrect from any sort of flight mode setting.

DonLakeFlyer commented:
Also the frame is MAV_FRAME_GLOBAL. Two things there:
* The values in ORBIT_EXECUTION_STATUS are always ints so I would think the frame would be MAV_FRAME_GLOBAL_INT.
* GCS tends to work in relative altitudes. Seems like returning MAV_FRAME_GLOBAL_RELATIVE_ALT_INT would be better.

DonLakeFlyer commented:
Also passing a -radius to DO_ORBIT does not make the orbit go counter-clockwise

DonLakeFlyer commented:
I've only look at telemetry from inspector so far. Would be good to get this merged in. And then I can hook up to QGC and I can merge than. Then we have something we can test end to end. Harder to do things in non-checked in branches on both sides.

MaEtUgR commented:
> I don't see an Orbit flight mode in px4_custom_mode.h which I can translate in QGC?

True, that's because I honestly forgot about that.

> When I go into Orbit it reports HEARTBEAT.custom_mode == 0 which is incorrect from any sort of flight mode setting.

😕 I have no idea why that should happen, maybe the same reason like above. I need to look into that tomorrow.

> Also the frame is MAV_FRAME_GLOBAL. Two things there:

My choice there was quite simple, I checked what I get from QGC via command in terms of frame and value and made sure to fill the exact same specs. All of this can be changed but I think we should try to have it consistent between command and telemetry if there's no obvious reason against that.

> Also passing a -radius to DO_ORBIT does not make the orbit go counter-clockwise

Sounds like logic is broken, I thought I iplemented correctly but because I didn't find a way to send a negative radius in QGC I never tested that.

> Would be good to get this merged in.

I'm fine with that, I'll fix CI and get a review. Then we can in my eyes merge and iterate with your feedback.

DonLakeFlyer commented:
@MaEtUgR I was incorrectly thinking that I would need this pull to be merged before I could push through telemetry support in QGC. I was wrong. First pull for orbit visuals from telemetry just got put up. Should be in daily in couple hours.

potaito commented:
> Another thing. Orbit telemetry comes across 10Hz. That seems pretty fast for something that in most normal cases isn't going to change values. I'd think even half of that would work fine.

Oh yeah, I initially even set it to 1Hz. :+1:

MaEtUgR commented:
> Also passing a -radius to DO_ORBIT does not make the orbit go counter-clockwise

You were right, I tested it but not the case where you only apply a new radius. Fixed in https://github.com/PX4/Firmware/pull/10989/commits/9f2c304897cc0d53f0d7f218e25ceb9249fbd9a4

> Another thing. Orbit telemetry comes across 10Hz.

I chose 10Hz because of possible dynamic changes like making adjustments with rc input or a beacon target as center. In these cases the pilot want's to see in realtime what's going on on the map to have feedback for corrections. I switched to 5Hz: https://github.com/PX4/Firmware/pull/10989/commits/5de33433ac58786701be5c25147bf62c70ae00d9 Remeber it's only sent during orbit.

MaEtUgR commented:
rebased on master after merging the base branch #10907

potaito commented:
Go to sleep 😅

MaEtUgR commented:
@DonLakeFlyer With latest QGC daily (after https://github.com/mavlink/qgroundcontrol/pull/7073) everything works fine 🎉  except for the mode not getting displayed as "Orbit" but rather "Unkown 217:33751040". In https://github.com/PX4/Firmware/pull/10989/commits/93e15f62bbff5697788ea036643edbfa539afdd5 you can see that I added the orbit as new submode. In the heartbeat `custom_mode` should now be reported like defined in the header. At least it's not 0 aynmore.

DonLakeFlyer commented:
> except for the mode not getting displayed as "Orbit" but rather "Unkown 217:33751040"

Yeah, now that I have that header file I can fix that.

DonLakeFlyer commented on src/modules/commander/px4_custom_mode.h:
```diff
@@ -66,6 +66,11 @@ enum PX4_CUSTOM_SUB_MODE_AUTO {
 };
 
+enum PX4_CUSTOM_SUB_MODE_POSCTL {
+	PX4_CUSTOM_SUB_MODE_POSCTL_POSCTL = 1,
```
I don't think this change is backward compatible. Previously there were no sub modes for POSCTL hence a POSCTL flight mode was indicated by main_mode = PX4_CUSTOM_MAIN_MODE_POSCTL and sub_mode = 0. This new header seems to change this such that now POSCTL is identified by sub_mode = 1.

Shouldn't PX4_CUSTOM_SUB_MODE_POSCTL_POSCTL = 0?

DonLakeFlyer replied:
@MaEtUgR FYI

DonLakeFlyer replied:
So I verified using this pull and QGC that the following is what is sent:
POSCTL - sub_mode = 0
ORBIT - sub_mode = 2

Which is good from a back compat standpoint but not what the headers says. I'm going to adjust the QGC header to this:
```
enum PX4_CUSTOM_SUB_MODE_POSCTL {
    PX4_CUSTOM_SUB_MODE_POSCTL_POSCTL = 0,
    PX4_CUSTOM_SUB_MODE_POSCTL_ORBIT
};
```

Under the assumption that is what the firmware will do as well.

MaEtUgR replied:
Sure, let's do that. Sorry I didn't pay enough attention to the numbers because I thought it's a shared header anyways but backwards compatibility is of course mandatory in this case. I'm glad if it works without adding a main mode (possible as last resort).

MaEtUgR replied:
@DonLakeFlyer I force-pushed `PX4_CUSTOM_SUB_MODE_POSCTL_POSCTL = 0`

MaEtUgR commented:
This works fine now according to all my tests.

**Off topic**
I learned in the process that the MAVLink message definitions apparently need to be up to date on each and every component redirecting the packets. Every relay parses all the messages completely and discards them if they are not known... in my opinion a flaw.
E.g. for HITL I had to update the xml in https://github.com/PX4/jMAVSim/tree/master/mavlink/message_definitions just because jmavsim relays the packets to QGC which is not something obvious (thanks @julianoes for the hint).

hamishwillee commented:
> Every relay parses all the messages completely and discards them if they are not known... in my opinion a flaw.

According to the MAVLink specification those systems are non-compliant. That can be inferred from the docs, but I made it explicit in https://github.com/mavlink/mavlink-devguide/pull/154.

Can you post defect reports against those systems - you can point them to the note at the end of https://mavlink.io/en/guide/routing.html#overview
